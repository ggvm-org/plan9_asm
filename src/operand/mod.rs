use std::fmt;

use crate::register_with_offset::Register;

use super::register_with_offset::RegisterWithOffset;
#[derive(Debug, PartialEq)]
pub enum Operand {
    Ident(String),
    Int(i64),
    RegisterWithOffset(RegisterWithOffset),
}

impl fmt::Display for Operand {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            Operand::Ident(s) => s.clone(),
            Operand::Int(n) => format!("${n}"),
            Operand::RegisterWithOffset(inner) => inner.to_string(),
        };
        write!(f, "{s}")
    }
}

macro_rules! impl_from_operand {
    ($from_ty:ty => Ident) => {
        impl From<$from_ty> for Operand {
            fn from(v: $from_ty) -> Self {
                Self::Ident(v.to_string())
            }
        }
    };
    ($from_ty:ty => RegisterWithOffset) => {
        impl From<$from_ty> for Operand {
            fn from(v: $from_ty) -> Self {
                Self::RegisterWithOffset(crate::register_with_offset::RegisterWithOffset::from(v))
            }
        }
    };
    ($from_ty:ty => $variant:ident) => {
        impl From<$from_ty> for Operand {
            fn from(v: $from_ty) -> Self {
                Self::$variant(v)
            }
        }
    };
}

impl_from_operand!(i64 => Int);
impl_from_operand!(&str => Ident);
impl_from_operand!(String => Ident);
impl_from_operand!(RegisterWithOffset => RegisterWithOffset);
impl_from_operand!(Register => RegisterWithOffset);

#[macro_export(local_inner_macros)]
macro_rules! operand {
    ($offset:expr => $register_variant:ident) => {
        $crate::operand::Operand::RegisterWithOffset(crate::register_with_offset!($offset => $register_variant))
    };
    ($register:ident) => {
       $crate::operand::Operand::RegisterWithOffset(crate::register_with_offset!($register))
    };
    ($expr:expr) => {
        $crate::operand::Operand::from($expr)
    };
}

snapshot_test!(operand: operand!(AX), operand!(16=>SP), operand!(1));
