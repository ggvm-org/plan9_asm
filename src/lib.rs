use std::fmt;

pub use self::jmp_target::JmpTarget;
#[macro_use]
mod common;

#[macro_use]
pub mod operand;
#[macro_use]
pub mod register_with_offset;

#[macro_use]
pub mod macros;

use operand::Operand;

#[macro_use]
pub mod jmp_target;

pub use operand::*;
pub use register_with_offset::*;
pub use Directive::{Nop, Ret};

#[derive(Debug, PartialEq)]
pub enum Directive {
    Text { package: String, name: String },
    Subq(Operand, Operand),
    Leaq(Operand, Operand),
    Movq(Operand, Operand),
    Call { package: String, name: String },
    Addq(Operand, Operand),
    Ret,

    // CMPQ	SP, 16(R14)
    Cmpq(Operand, Operand),
    // PCDATA	$0, $-2
    PCData(Operand, Operand),

    // epi:
    Label(String),
    // NOP
    Nop,
    // JMP body
    Jmp(JmpTarget),
    // Jls	epi
    Jls(JmpTarget),
}

impl fmt::Display for Directive {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            Self::Text { package, name } => format!("TEXT	{}.{}(SB), 4, $0-0", package, name),
            Self::Subq(left, right) => format!("SUBQ	{}, {}", left, right),
            Self::Call { package, name } => format!("CALL    {package}Â·{name}(SB)"),
            Self::Addq(left, right) => format!("ADDQ	{}, {}", left, right),
            Self::Movq(left, right) => format!("MOVQ	{}, {}", left, right),
            Self::Cmpq(left, right) => format!("CMPQ	{}, {}", left, right),
            Self::Jls(target) => format!("JLS	{}", target),
            Self::PCData(left, right) => format!("PCDATA {}, {}", left, right),
            Self::Label(label_name) => format!("{}:", label_name),
            Self::Jmp(target) => format!("JMP	{}", target),
            Self::Nop => "NOP".to_string(),
            _ => unimplemented!(),
        };
        write!(f, "{s}")
    }
}

impl fmt::Display for JmpTarget {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            Self::Addr(s) => s.to_string(),
            Self::Label(l) => l.to_string(),
        };
        write!(f, "{s}")
    }
}
