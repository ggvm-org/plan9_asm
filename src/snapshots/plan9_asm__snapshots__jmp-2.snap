---
source: src/lib.rs
assertion_line: 253
expression: "JMP!(\"epi\")"

---
JMP	epi
